crates/red_knot_python_semantic/src/types/narrow.rs

```diff
@@ -194,6 +201,52 @@ impl<'db> NarrowingConstraintsBuilder<'db> {
         }
     }
 
+    /// Generate a constraint from the *type* of the second argument of an `isinstance` call.
+    ///
+    /// Example: for `isinstance(…, str)`, we would infer `Type::ClassLiteral(str)` from the
+    /// second argument, but we need to generate a `Type::Instance(str)` constraint that can
+    /// be used to narrow down the type of the first argument.
+    fn to_isinstance_constraint(&self, classinfo: &Type<'db>) -> Option<Type<'db>> {
```

Comment:
nit: not sure about the name of this method. I think the usual convention is that `to_*` methods convert `self` into another type -- but here we're converting a passed-in argument to another type.

This routine actually looks like it could either be a method on `Type` or a freestanding function that accepts `db` and `classinfo`

Reply:
> This routine actually looks like it could either be a method on `Type` 

I don't know. `Type` seems too prominent a place for this helper function.

> or a freestanding function that accepts `db` and `classinfo`

Yeah, okay.

> nit: not sure about the name of this method. I think the usual convention is that `to_*` methods convert `self` into another type -- but here we're converting a passed-in argument to another type.

It's now a free function, but I renamed it to `generate_isinstance_constraint` anyway.

Reply:
Yeah, I like what you've done best, I think. Having it in `narrow.rs` improves encapsulation -- we should only need this functionality for narrowing, not for anything else -- so it's better not to have it as a method on `Type`